Building GraphRAG Python MCP tools (#368)
neo4j-graphacademy/courses: asciidoc/courses/genai-mcp-build-custom-tools-python/modules/2-database-features/lessons/1-connecting-neo4j/lesson.adoc

----
from neo4j import GraphDatabase

@mcp.tool()
def get_movies() -> list[dict]:
    """Get a list of movies"""
    # Creating a new driver for every tool call!
    driver = GraphDatabase.driver(uri, auth=(user, password))
    with driver.session() as session:
        result = session.run("MATCH (m:Movie) RETURN m LIMIT 10")
        return [record.data() for record in result]
    driver.close()
----

This approach has several problems:

* **Performance**: Creating a new driver connection for every tool call is slow and inefficient
* **Resource leaks**: If the tool fails, the driver may not be closed properly
* **Connection pooling**: Neo4j drivers maintain connection pools that should be reused across requests
* **Best practices**: The Neo4j driver should be created once and closed when the server shuts down

== Introducing Lifespan Management

FastMCP provides a **lifespan** feature that allows you to:

1. **Initialize resources** when the server starts (e.g., create database connections)
2. **Clean up resources** when the server shuts down (e.g., close connections)
3. **Share resources** across all tools and resources in your server

=== The Lifespan Context Manager

> To share objects across the server, we can create a function that initializes resources when the server starts and cleans them up when it shuts down.

Reviewer: ```suggestion
To share objects across the server, you create a function that initializes resources when the server starts and cleans them up when it shuts down.
```

> The function yields an object that contains the FastMCP server will make available to any tools and resources that request it.

Reviewer: ```suggestion
The function yields an object that contains the FastMCP server that it will make available to any tools and resources that request it.
```

Let's take a look at a full example:

[source,python]
.Full lifespan example
----
import os
from collections.abc import AsyncIterator
from contextlib import asynccontextmanager
from dataclasses import dataclass

from neo4j import AsyncGraphDatabase, AsyncDriver
from mcp.server.fastmcp import Context, FastMCP

# 1. Define a context class to hold your resources
@dataclass
class AppContext:
    """Application context with shared resources."""
    driver: AsyncDriver
    database: str

# 2. Create the lifespan context manager
@asynccontextmanager
async def app_lifespan(server: FastMCP) -> AsyncIterator[AppContext]:
    """Manage application lifecycle."""
    
    # Startup: Read credentials from environment variables

>     uri = os.getenv("NEO4J_URI", "bolt://localhost:7687")

Reviewer: Maybe remove the defaults?! If they dont exist it should fail.

----

The context class holds all resources that should be shared across your server.
Using a dataclass with type hints provides better IDE support and type safety.

For your MCP server, you will need to define a context class that holds the Neo4j driver and the current database.

==== 2. Create the Lifespan Context Manager

[source,python]
.Context manager function
----
@asynccontextmanager
async def app_lifespan(server: FastMCP) -> AsyncIterator[AppContext]:
----

The `@asynccontextmanager` decorator creates an async context manager.
Code before `yield` runs at **server startup**, code in `finally` runs at **server shutdown**.

==== 3. Use Environment Variables

[source,python]
.Accessing environment variables
----
uri = os.getenv("NEO4J_URI", "bolt://localhost:7687")
username = os.getenv("NEO4J_USERNAME", "neo4j")
password = os.getenv("NEO4J_PASSWORD", "password")
database = os.getenv("NEO4J_DATABASE", "neo4j")
----

The credentials needed to connect to the database are read from environment variables.

[WARNING]
.Never hardcode credentials
====
Environment variables contain sensitive information and allow different configurations for development and production.

> They should never be written into your code.

Reviewer: Yep, we should remove the default values.
